ddsrecorder_participants/src/cpp/recorder/size_tracker/McapSizeTracker.cpp

```diff
@@ -0,0 +1,192 @@
+        const mcap::Schema& schema)
+{
+    constexpr std::uint64_t NUMBER_OF_TIMES_COPIED = 2;
+    constexpr std::uint64_t CONST_SCHEMA = 5;
+
+    std::uint64_t size = MCAP_SCHEMAS_OVERHEAD;
+    size += schema.name.size();
+    size += schema.encoding.size();
+    size += schema.data.size();
+    size *= NUMBER_OF_TIMES_COPIED;
+
+    size -= CONST_SCHEMA;
+
+    return size;
+}
+
+std::uint64_t McapSizeTracker::get_blank_schema_size(
+        const bool& ros2_types,
+        const std::string& schema_name)
+{
+    constexpr int NUMBER_OF_TIMES_COPIED = 2;
+    constexpr std::uint64_t CONST_SCHEMA = 5;
+
+    std::string encoding = ros2_types ? "ros2msg" : "omgidl";
+
+    std::uint64_t size = MCAP_SCHEMAS_OVERHEAD;
+    size += schema_name.size();
+    size += encoding.size();
+    size *= NUMBER_OF_TIMES_COPIED;
+    size -= CONST_SCHEMA;
+
+    return size;
+}
+
+std::uint64_t McapSizeTracker::get_channel_size(
+        const mcap::Channel& channel,
+        const uint64_t& size_metadata)
+{
+    constexpr int NUMBER_OF_TIMES_COPIED = 2;
+
+    std::uint64_t size = MCAP_CHANNEL_OVERHEAD;
+    size += channel.topic.size();
+    size += channel.messageEncoding.size();
+    size += size_metadata;
+    size *= NUMBER_OF_TIMES_COPIED;
+
+    return size;
+}
+
+std::uint64_t McapSizeTracker::get_blank_channel_size(
+        const std::string& channel_name)
+{
+    constexpr std::uint64_t FIXED_SIZE = MCAP_CHANNEL_OVERHEAD + 103; //Encoding: 3 + Metadata:100
+    constexpr int NUMBER_OF_TIMES_COPIED = 2;
+
+    std::uint64_t size = FIXED_SIZE;
+    size += channel_name.size();
+    size *= NUMBER_OF_TIMES_COPIED;
+
+    return size;
+}
+
+std::uint64_t McapSizeTracker::get_attachment_size(
+        const std::unique_ptr<fastrtps::rtps::SerializedPayload_t>& dynamic_attachment_serialized_payload)
+{
+    constexpr std::uint64_t NUMBER_OF_TIMES_COPIED = 1;
+
+    std::uint64_t size = MCAP_ATTACHMENT_OVERHEAD;
+    size += dynamic_attachment_serialized_payload->length;
+    size *= NUMBER_OF_TIMES_COPIED;
+
+    return size;
+}
+
+void McapSizeTracker::decrease_mcap_size(
+        const std::uint64_t& size)
+{
+    assert(mcap_size_ >= size);
+    mcap_size_ -= size;
+}
+
+void McapSizeTracker::check_and_increase_mcap_size_final(
+        const std::uint64_t& size)
+{
+    if (!disk_full_)
+    {
+        if ((mcap_size_ + size) > space_available_when_open_)
+        {
+            disk_full_ = true;
+            throw std::overflow_error(
+                        STR_ENTRY << "Attempted to write an MCAP of size: " << mcap_size_ <<
+                    ", but there is not enough space available on disk: " << space_available_when_open_);
+        }
+        else
+        {
+            mcap_size_ += size;
+        }
+    }
+    else
+    {
+        logInfo(DDSRECORDER_MCAP_HANDLER,
+                "Disk is full, adding data previously taken into account");
+        assert((mcap_size_ + size) <= space_available_when_open_);
```

Comment:
I'm not sure this condition makes sense. I believe after the disk is set as full, there should no longer be a call to this method, so perhaps an assert false + log error could be added instead.

Reply:
What I believe should be added is also an attribute tracking the "actual" (predicted) written size, and so in this case we should assert it never surpasses the threshold.